android/src/main/kotlin/com/fluttercandies/photo_manager/permission/PermissionDelegate.kt

```diff
@@ -0,0 +1,175 @@
+     */
+    fun havePermissions(context: Context, vararg permission: String): Boolean {
+        val result = permission.all { havePermission(context, it) }
+        LogUtils.debug("[$tag] havePermissions: ${permission.toList()}, result: $result")
+        return result
+    }
+
+    /**
+     * Request permission.
+     *
+     * The [permissionsUtils] is used to get the activity.
+     * The [context] is used to check the permission is in the manifest.
+     * The [requestType] is passed from the dart code.
+     * The [mediaLocation] is passed from the dart code.
+     */
+    abstract fun requestPermission(
+        permissionsUtils: PermissionsUtils,
+        context: Context,
+        requestType: Int,
+        mediaLocation: Boolean,
+    )
+
+    /**
+     * Check if the [requestType] is granted for the user and it is in the manifest.
+     */
+    abstract fun havePermissions(context: Context, requestType: Int): Boolean
+
+    /**
+     * Check if the [mediaLocation] is granted for the user and it is in the manifest.
+     */
+    abstract fun haveMediaLocation(context: Context): Boolean
+
+    companion object {
+        const val requestCode = 3001
+        const val limitedRequestCode = 3002
+
+        /**
+         * Create a [PermissionDelegate] by the sdk version.
+         */
+        fun create(): PermissionDelegate {
+            return when (Build.VERSION.SDK_INT) {
+                in 1 until 23 -> PermissionDelegate19()
+                in 23 until 29 -> PermissionDelegate23()
+                29 -> PermissionDelegate29()
+                in 30 until 33 -> PermissionDelegate30()
+                33 -> PermissionDelegate33()
+                in 34 until Int.MAX_VALUE -> PermissionDelegate34()
+                else -> throw UnsupportedOperationException(
+                    "This sdk version is not supported yet."
+                )
+            }
+        }
+    }
+
+    open fun isHandlePermissionResult(): Boolean {
+        return false
+    }
+
+    open fun handlePermissionResult(
+        permissionsUtils: PermissionsUtils,
+        context: Context,
+        permissions: Array<String>,
+        grantResults: IntArray,
+        needToRequestPermissionsList: MutableList<String>,
+        deniedPermissionsList: MutableList<String>,
+        grantedPermissionsList: MutableList<String>,
+        requestCode: Int
+    ) {
+        throw UnsupportedOperationException(
+            "handlePermissionResult is not implemented," +
+                    " please implement it in your delegate."
+        )
+    }
+
+    open fun presentLimited(
+        permissionsUtils: PermissionsUtils,
+        context: Application,
+        type: Int,
+        resultHandler: ResultHandler
+    ) {
+        LogUtils.debug("[$tag] presentLimited is not implemented")
+        resultHandler.reply(null)
+    }
+
+    abstract fun getAuthValue(context: Application, requestType: Int, mediaLocation: Boolean): PermissionResult
+
+}
```

Comment:
Need a proper EOF (So as other files).

Reply:
I use [code style migration](https://kotlinlang.org/docs/code-style-migration-guide.html#migration-to-a-new-code-style-discussion) to reformat all kotlin code files.

Kotlin does not seem to have this style requirement similar to Dart's EOF.